subnet: add infrastructure and tests for network watches

dcbw opened the pull request:
This PR depends on https://github.com/coreos/flannel/pull/282 (which is the first commit here too) so that one should get sorted out first.

dcbw commented:
I'm not quite sure why the Travis build is failing, I cannot reproduce on two different machines with either Go 1.4.2 or Go 1.3.3... @rajatchopra any thoughts?  In the remote test on the Travis build, client.go::doTestWatchNetworks() times out waiting to receive the NetworkAdded event from the server.

dcbw commented:
[test]

eyakubovich commented:
I really like the functionality, thanks for adding this.

dcbw commented:
@eyakubovich ok I'll rework to separate the watch structs.  Still chasing down why the Travis tests are failing too.

dcbw commented:
Ignore any asdfasdfadsf commits too, that's just my sad attempt to figure out why the Travis builds are failing.

dcbw commented:
@eyakubovich are the EventType strings part of the API/ABI too?  Do I need to keep those backwards compatible?

eyakubovich commented:
@dcbw Yes, they're part of the API. Granted, I don't think too many people use this remote functionality yet so I don't think it's a big deal to break the API even if we don't bump the version. However in this case I don't think there's a reason to do so.

dcbw commented:
@eyakubovich I've reverted the EventType API string changes and fixed up the failing Travis test.  I think this branch is ready for more review.  Thanks!

eyakubovich commented on remote/client.go:
```diff
@@ -178,19 +178,50 @@ func (m *RemoteManager) WatchLeases(ctx context.Context, network string, cursor
+
+func (m *RemoteManager) WatchNetworks(ctx context.Context, cursor interface{}) (subnet.NetworkWatchResult, error) {
+	url := m.base + "/"
+	if cursor != nil {
```
Can you factor the common code out into something like:

```
func (m *RemoteManager) watch(ctx context.Context, cursor interface{}, url string, wr {}interface) error {
    if cursor != nil {
        c, ok := cursor.(string)
        if !ok {
            return subnet.NetworkWatchResult{}, fmt.Errorf("internal error: RemoteManager.watch received non-string cursor")
        }

        url = fmt.Sprintf("%v?next=%v", url, c)
    }

    resp, err := m.httpGet(ctx, url)
    if err != nil {
        return err
    }

    if resp.StatusCode != http.StatusOK {
        return httpError(resp)
    }

    if err := json.NewDecoder(resp.Body).Decode(wr); err != nil {
        return err
    }
    if _, ok := wr.Cursor.(string); !ok {
        return subnet.NetworkWatchResult{}, fmt.Errorf("watch returned non-string cursor")
    }
}
```

dcbw replied:
> Can you factor the common code out into something like:

Done.

eyakubovich commented on remote/remote_test.go:
```diff
@@ -168,3 +171,84 @@ func doTestWatch(t *testing.T, sm subnet.Manager) {
+	<-events
+
+	expectedNetname := "foobar"
+	go func() {
```
Why a separate goroutine? The current goroutine is just going to block on line 214 anyway.

dcbw replied:
> Why a separate goroutine? The current goroutine is just going to block on line 214 anyway.

This got removed in some later cleanups.

eyakubovich commented on remote/remote_test.go:
```diff
@@ -168,3 +171,84 @@ func doTestWatch(t *testing.T, sm subnet.Manager) {
+		// Add a new network a short time later to ensure that the
+		// client-side network watch triggers a server watch before the
+		// network is added directly in the server with CreateNetwork()
+		time.Sleep(time.Second)
```
I don't think I fully understand the need for sleep. Is the desire to exercise the code where the server does an etcd watch prior to the network key being created? In that case I guess it's ok but the code should work regardless. That is, this also needs to work:

1. Client side issues a watch
2. A new network is created in etcd
3. The server receives the watch and processes it by immediately returning
4. The client gets notified of the new network

dcbw replied:
> I don't think I fully understand the need for sleep. Is the desire to exercise the code where the server does an etcd watch prior to the network key being created?

The server needs to be watching before the new network is created on the client side, otherwise the server-side will completely miss the event.  So subnet.WatchNetworks() must be running and sitting in RemoteManager::httpDo() (which triggers the server's handleWatchNetworks() function -> etcd.go::WatchNetworks()) before we execute the CreateNetwork() call.  But since subnet.WatchNetworks() is done in a goroutine and the actual HTTP request that triggers the server's handleWatchNetworks is way down the callstack, there's no way to guarantee that this will all happen because they are concurrent.

Obviously this is only a problem in the testcases, because normal clients will never be creating networks.

So possibly my approach of directly modifying the server's registry to add the network and generate the event isn't the right way to go about this, because of this problem...

eyakubovich replied:
Your approach of directly modifying the server's register is just fine -- you're just simulating an external event.

> The server needs to be watching before the new network is created on the client side, otherwise the server-side will completely miss the event.

This is the scary statement. `subnet.WatchNetworks` internally maintains a cursor so it should not miss an event. If `CreateNetwork` executes before `etcd.go::WatchNetworks()`, that should be fine. When it does get to `etcd.go::WatchNetworks()`, the cursor will be pointing slightly behind the "end" and it will immediately return the added network. That is the whole point of maintaining and propagating the cursor.

dcbw replied:
> This is the scary statement.

Fair enough; I'll figure out why the cursor isn't doing what it should be doing here.

eyakubovich commented on remote/remote_test.go:
```diff
@@ -168,3 +171,84 @@ func doTestWatch(t *testing.T, sm subnet.Manager) {
+		t.Errorf("WatchNetwork create produced wrong network: expected %s, got %s", expectedNetname, evt.Network)
+	}
+
+	go func() {
```
Same as comments above

eyakubovich commented on subnet/subnet.go:
```diff
@@ -44,26 +44,37 @@ type (
-		Lease Lease     `json:"lease"`
+		Type    EventType `json:"type"`
+		Lease   Lease     `json:"lease"`
+		Network string    `json:"network"`
```
Since we're using the same Event struct for both subnet and lease events (which is fine), add `omitempty` so it's clean on the wire.

dcbw replied:
> add omitempty so it's clean on the wire.

Done

eyakubovich commented on subnet/subnet.go:
```diff
@@ -44,26 +44,37 @@ type (
 const (
 	SubnetAdded EventType = iota
 	SubnetRemoved
+	NetworkAdded
```
Let's just have two events: `Added` and `Removed`. Or maybe call them `EventAdded` and `EventRemoved`. But they will be shared between both the lease and network events. There should be no confusion since the API endpoints and methods are different.

dcbw replied:
> Let's just have two events: Added and Removed. Or maybe call them EventAdded and EventRemoved.

Done.

dcbw commented:
I've got the stuff commented as "Done" done locally; I have to fix up one or two other things as a result of these, then figure out the cursor testcase issue, then I'll re-push, hopefully tomorrow.  Keep the comments coming if you're still reviewing, I'll get to them tomorrow or next week.  Thanks!

(my next steps are factoring out the network code into a network/manager.go that watches for networks coming and going, and creates or removes networks as needed, but that's another PR...)

dcbw commented:
@eyakubovich repushed; I believe all your comments have been addressed.  Thanks!

eyakubovich commented on remote/client.go:
```diff
@@ -164,13 +164,11 @@ func (m *RemoteManager) RenewLease(ctx context.Context, network string, lease *s
 		c, ok := cursor.(string)
 		if !ok {
-			return subnet.WatchResult{}, fmt.Errorf("internal error: RemoteManager.WatchLeases received non-string cursor")
+			return fmt.Errorf("internal error: RemoteManager.WatchLeases received non-string cursor")
```
`RemoteManager.WatchLeases` -> `RemoteManager.watch`

dcbw replied:
Done in latest push.

eyakubovich commented on subnet/etcd.go:
```diff
@@ -347,36 +395,118 @@ func parseSubnetWatchResponse(resp *etcd.Response) (WatchResult, error) {
+	return wr, nil
+}
+
+// leaseWatchReset is called when incremental network watch failed and we need to grab a snapshot
```
`leaseWatchReset` -> `networkWatchReset`

dcbw replied:
Done in latest push.

eyakubovich commented on subnet/registry.go:
```diff
@@ -131,13 +132,19 @@ func (esr *etcdSubnetRegistry) deleteSubnet(ctx context.Context, network, sn str
+}
+
+func (esr *etcdSubnetRegistry) getNetworks(ctx context.Context) (*etcd.Response, error) {
+	key := path.Join(esr.etcdCfg.Prefix, "network")
```
What is this "network"? The current struct is:
```
- etcd-prefix
     |- mynetwork1
           | - config
           | - subnets
                  |- mysubnet1
                  |- mysubnet2
     |- mynetwork2
           | - config
           | - subnets
                  |- mysubnet1
                  |- mysubnet2
```

dcbw replied:
Yeah, just found that out when refactoring the network/ code to watch networks.  It's bogus; removed.

eyakubovich commented on subnet/subnet.go:
```diff
@@ -96,5 +105,7 @@ type Manager interface {
 	RenewLease(ctx context.Context, network string, lease *Lease) error
-	WatchLeases(ctx context.Context, network string, cursor interface{}) (WatchResult, error)
+	WatchLeases(ctx context.Context, network string, cursor interface{}) (LeaseWatchResult, error)
+	GetNetworks(ctx context.Context) ([]string, error)
```
Let's combine `GetNetworks` and `WatchNetworks` into one. `GetNetworks` is just a watch with `nil` cursor. I'm actually working on another feature and doing exactly that for leases. We can rename `WatchNetworks` to `GetNetworks` if you think it's better.

dcbw replied:
Done in the latest push.  I opted for "WatchNetworks" instead of GetNetworks, but could go either way if you really prefer GetNetworks.

dcbw commented:
Repushed with fixes for all your previous comments.

eyakubovich commented on remote/client.go:
```diff
@@ -178,19 +176,49 @@ func (m *RemoteManager) WatchLeases(ctx context.Context, network string, cursor
+	}
+
+	// Only return a cursor for the 'watch' operation
+	if cursor != nil {
```
Sorry, I think you misunderstood me a bit. Yes, I proposed combining watch+get much like you've done here. However you should always return a cursor. Otherwise the watch won't work -- you kick the watch off by getting all the networks (no cursor passed in) and then watch for increments (using cursor) if you so desire. I think what you have for `WatchLeases` is perfect.

eyakubovich commented on remote/server.go:
```diff
@@ -152,6 +152,38 @@ func handleWatchLeases(ctx context.Context, sm subnet.Manager, w http.ResponseWr
+	}
+
+	// Only return a cursor for the 'watch' operation
+	if cursor != nil {
```
Always return a cursor

eyakubovich commented on subnet/etcd.go:
```diff
@@ -347,36 +390,118 @@ func parseSubnetWatchResponse(resp *etcd.Response) (WatchResult, error) {
+}
+
+func (m *EtcdManager) parseNetworkWatchResponse(resp *etcd.Response) (NetworkWatchResult, error) {
+	netname, err := m.parseNetworkKey(resp.Node.Key)
```
The watch is recursive on something like `/coreos.com/network/`. Wouldn't then this fire on every key change, like subnets getting added? In that case `ParseConfig` will fail and that error propagated.

dcbw replied:
> The watch is recursive on something like /coreos.com/network/. Wouldn't then this fire on every key change, like subnets getting added? In that case ParseConfig will fail and that error propagated.

Yes, it should be recursive, since we consider the 'config' key authoritative; if it's missing the network is dead to us.  I've fixed that up now to ignore sub-keys and not return an error.

eyakubovich commented:
@dcbw I know it's a long review but I think this PR getting very close.

dcbw commented:
> I know it's a long review but I think this PR getting very close.

Yeah, I've been trying to track down the remote testcase hang before replying, not there yet.  Appears to happen about 1/5 the time on the Travis CI machines:

E0818 18:58:20.993249 02590 mock_registry.go:286] #### mockreg::CreateNetwork() sent add event; events (chan *client.Response)(0xc2080a2000)
E0818 18:58:20.993737 02590 remote_test.go:202] #### test::doTestWatchNetworks() added network
E0818 18:58:20.993757 02590 remote_test.go:207] #### test::doTestWatchNetworks() waiting for Add event
E0818 18:58:20.993922 02590 server.go:189] #### server::handleNetworks() start cursor "4"
E0818 18:58:20.993731 02590 mock_registry.go:188] #### mockreg::watch() got event &client.Response{Action:"add", Node:(*client.Node)(0xc208004ea0), PrevNode:(*client.Node)(nil), Index:0x0} modidx 5 since 4
E0818 18:58:20.993937 02590 etcd.go:335] #### etcd::WatchNetworks() start cursor "4"
E0818 18:58:20.993963 02590 mock_registry.go:193] #### mockreg::watch() returning event &client.Response{Action:"add", Node:(*client.Node)(0xc208004ea0), PrevNode:(*client.Node)(nil), Index:0x0}
E0818 18:58:20.993994 02590 etcd.go:346] #### etcd::WatchNetworks() about to watch networks nextIndex 0x4
E0818 18:58:20.994034 02590 mock_registry.go:180] #### mockreg::watch() starting watch net  since 4

Since the server-side misses the "add" event it just sits in etcd::WatchNetworks()/mockreg::watch() waiting for something to happen, but nothing does.  The root cause appears to be something still sitting in the MockRegistry's watch() method, which is grabbing the event from the queue before the server starts watching.

Say the cursor gets updated internally to "5" in response to some random event.  Then a client watches networks with an older cursor "4", and we get to server.go::handleNetworks().  That calls into etcd.go::WatchNetworks() and then to mock_registry.go::watch() with cursor "4".  Which just sits there waiting for some new event which (in the case of the remote tests) never comes.

Shouldn't something on the server-side return a snapshot when the client-provided cursor is less than the registry cursor?  Something like this perhaps?

func (msr *MockSubnetRegistry) watch(ctx context.Context, network string, since uint64) (*etcd.Response, error) {
	for {
+		if since < msr.index {
+			return nil, etcd.Error{
+				Code: etcd.ErrorCodeEventIndexCleared,
+				Cause: "out of date",
+				Message: "cursor is out of date",
+				Index: msr.index,
+			}
+		}

		select {
		case <-ctx.Done():

dcbw commented:
> Shouldn't something on the server-side return a snapshot when the client-provided cursor is less than the registry cursor? Something like this perhaps?

So I added this and pushed it, and it seems to pass the testcases in Travis now much more reliably.  It seems like the correct thing to do, but your input would be appreciated here.  Thanks!

Repushed, hopefully addressing all your previous comments.

eyakubovich commented:
LGTM. Thank you for your work.
